cohorts/2023/02-experiment-tracking/homework/datahandler.py

```diff
@@ -0,0 +1,196 @@
+        self.s3_url = s3_url
+
+    def download_data(self, parquet_file: str) -> Path:
+        """Download a data file from the specified S3 bucket.
+
+        Args:
+            parquet_file (str): The name of the data file to download.
+
+        Returns:
+            Path: The path to the downloaded file.
+        """
+
+        file_path = self.data_dir / parquet_file
+        url = self.s3_url + parquet_file
+        logger.info("Attempting to download filename %s", url)
+        if not os.path.isfile(file_path):
+            logger.info("File does not exist, downloading from S3 bucket.")
+            if not os.path.exists(os.path.dirname(file_path)):
+                os.makedirs(os.path.dirname(file_path))
+            subprocess.run(["wget", "-O", file_path, url], check=True)
+            logger.info("File downloaded successfully and saved at %s", file_path)
+        else:
+            logger.info("File already exists.")
+        return file_path
+
+    def preprocess_dataset(self, dataset: pd.DataFrame) -> pd.DataFrame:
+        """
+        Preprocesses the input dataset by adding new features and filtering rows.
+
+        This method performs preprocessing operations on the input dataset to enhance its
+        suitability for downstream tasks. It calculates the duration of each trip, filters out
+        rows with unrealistic durations, and creates a concatenated feature representing the
+        pickup and dropoff locations. Additionally, it converts specific columns to string
+        data type for categorical analysis.
+
+        Args:
+            dataset (pd.DataFrame): The input dataset containing raw trip data.
+
+        Returns:
+            pd.DataFrame: A preprocessed dataset with added features and filtered rows.
+
+        Example:
+            # Load a raw dataset
+            raw_dataset = pd.read_parquet("raw_dataset.parquet")
+
+            # Create a DataHandler object
+            handler = DataHandler(data_dir=Path("data/"), s3_url="https://s3.example.com/")
+
+            # Preprocess the raw dataset
+            preprocessed_dataset = handler.preprocess_dataset(raw_dataset)
+        """
+
+        dataset["duration"] = (
+            dataset["lpep_dropoff_datetime"] - dataset["lpep_pickup_datetime"]
+        ).dt.total_seconds() / 60
+        filtered_dataset = dataset[
+            (dataset.duration >= 1) & (dataset.duration <= 60)
+        ].copy()
+        str_columns = ["PULocationID", "DOLocationID"]
+        filtered_dataset.loc[:, str_columns] = filtered_dataset[str_columns].astype(str)
+        filtered_dataset.loc[:, "PU_DO"] = (
+            filtered_dataset["PULocationID"] + "_" + filtered_dataset["DOLocationID"]
+        )
+        filtered_dataset.loc[:, "PU_DO"] = filtered_dataset["PU_DO"].astype(str)
+        categorical = ["PU_DO"]
+        numerical = ["trip_distance", "tip_amount"]
+
+        return filtered_dataset[categorical + numerical]
+
+    def load_dataset(
```

Comment:
Are these methods only called inside the object class? If so, it might be good to make them private.